### PR TITLE
[Security Fix] Upgrade mysql connector to 8.2.0

### PR DESCRIPTION
**Pull request by mustajibmk**
### Description

Upgrade MySql connector to 8.2.0 to fix security vulnerabilities.

#### Release note

<hr>

##### Key changed/added classes in this PR
 * Upgrade version
 * Replace deprecated classes with the new ones - https://dev.mysql.com/doc/connectors/en/connector-j-upgrading-to-8.0.html
 * The upgraded version of MySQL connector does not have the `parseURL` method in `com.mysql.jdbc.NonRegisteringDriver` previously used. Instead, we use the `com.mysql.cj.conf.ConnectionUrlParser` which only checks if the schema of the string matches the prescribed format. The test cases related to string parsing are hence removed.

<hr>

This PR has:

- [x] been self-reviewed.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java [hunk 1]
```diff
@@ -177,19 +178,26 @@ public static Set<String> tryParsePostgresConnectionUri(String connectionUri)
...
+    Map<String,String> propertyMap =  (Map<String, String>) getProperties.invoke(connectionUrlParser);
+    for (Map.Entry<String, String> entry : propertyMap.entrySet()) {
+      if (entry.getKey() != null) {
+        if(entry.getValue() == null) {
+          properties.setProperty(entry.getKey(), "");
+        }else {
+          properties.setProperty(entry.getKey(), entry.getValue());
+        }
+      }
+    }
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:199]: the code is not well formatted. You can import the checkstyle that's available in dev folder.

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java [hunk 2]
```diff
@@ -177,19 +178,26 @@ public static Set<String> tryParsePostgresConnectionUri(String connectionUri)
...
+    Class<?> connectionUrlParserClass = Class.forName(MYSQL_CJ_CONF_CONNECTION_URL_PARSER);
+    Method parseUrl = connectionUrlParserClass.getMethod("parseConnectionString", String.class);
+    // almost the same as postgres
+    Object connectionUrlParser = parseUrl.invoke(null, connectionUri);
+    Method getProperties = connectionUrlParserClass.getMethod("getProperties");
+    Properties properties = new Properties();
+    Map<String,String> propertyMap =  (Map<String, String>) getProperties.invoke(connectionUrlParser);
+    for (Map.Entry<String, String> entry : propertyMap.entrySet()) {
+      if (entry.getKey() != null) {
+        if(entry.getValue() == null) {
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:193]: is this actually possible?

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java [hunk 3]
```diff
@@ -177,19 +178,26 @@ public static Set<String> tryParsePostgresConnectionUri(String connectionUri)
...
+          properties.setProperty(entry.getKey(), "");
+        }else {
+          properties.setProperty(entry.getKey(), entry.getValue());
+        }
+      }
+    }
+    if (properties.isEmpty()) {
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:200]: why would you throw an error if there are no properties associated with the connection string? Instead you should throw an error if properties is null for some reason.

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java (hunk 2 above)
Comment (mustajibmk) [reply processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java]: Yes, it is possible. Like in this [case](https://github.com/apache/druid/blob/e373f6269251655f5be93ce895aee8dee8cc67dd/processing/src/test/java/org/apache/druid/utils/ConnectionUriUtilsTest.java#L42) where `keyonly` doesn't have any value associated with it.

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java (hunk 3 above)
Comment (mustajibmk) [reply processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:202]: With the initialisation of `properties`. It cannot be null anymore, hence empty remains the only possible scenario when nothing could be extracted from the connection string.

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java [hunk 4]
```diff
@@ -177,21 +182,75 @@ public static Set<String> tryParsePostgresConnectionUri(String connectionUri)
...
-    Properties properties = (Properties) parseUrl.invoke(
-        driverClass.getConstructor().newInstance(),
-        connectionUri,
-        null
-    );
+    Class<?> connectionUrlClass = Class.forName(MYSQL_CONNECTION_URL);
+    Class<?> hostInfoClass = Class.forName(MYSQL_HOST_INFO);
+    Class<?> urlParserClass = Class.forName(MYSQL_CONNECTION_URL_PARSER);
+    Method isConnectionStringSupported = urlParserClass.getMethod("isConnectionStringSupported", String.class);
+    boolean isConnectionUrlSupported = (boolean) isConnectionStringSupported.invoke(urlParserClass, connectionUri);
```
Comment (mustajibmk) [inline processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:192]: Given that `isConnectionStringSupported` only checks the schema which to a certain extends already happens [here](https://github.com/apache/druid/blob/13e16e36206fb26c0f32c954cfd641375f94e9f1/processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java#L43). Should we have this check in place?

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java [hunk 5]
```diff
@@ -177,21 +182,75 @@ public static Set<String> tryParsePostgresConnectionUri(String connectionUri)
...
   {
-    Class<?> driverClass = Class.forName(MYSQL_NON_REGISTERING_DRIVER);
-    Method parseUrl = driverClass.getMethod("parseURL", String.class, Properties.class);
-    // almost the same as postgres, but is an instance level method
-    Properties properties = (Properties) parseUrl.invoke(
-        driverClass.getConstructor().newInstance(),
-        connectionUri,
-        null
-    );
+    Class<?> connectionUrlClass = Class.forName(MYSQL_CONNECTION_URL);
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:188]: can you describe your logic in a bit more detail?

File: extensions-core/lookups-cached-global/src/test/java/org/apache/druid/query/lookup/namespace/JdbcExtractionNamespaceUrlCheckTest.java [hunk 6]
```diff
@@ -159,46 +159,6 @@ public boolean isEnforceAllowedProperties()
           }
       );
     }
-
-    @Test
-    public void testWhenInvalidUrlFormat()
```
[the remark this answers is not in the document]
Comment (mustajibmk) [reply extensions-core/lookups-cached-global/src/test/java/org/apache/druid/query/lookup/namespace/JdbcExtractionNamespaceUrlCheckTest.java:164]: The parser no longer checks if every component is in the expected format it only checks if the URL starts with one of the required schema i.e `jdbc:mysql`,`jdbc:mysql:loadbalancer`, etc. So we will not be able to determine if this is a valid/invalid URL just by using the parser. 

The way to determine if the URL is valid is by creating a connection with the DB.

File: processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java (hunk 5 above)
Comment (mustajibmk) [reply processing/src/main/java/org/apache/druid/utils/ConnectionUriUtils.java:188]: Now the method 
- Checks if the string is in acceptable format 
- Get the `list<hostInfo>` obtained from the string
- From the `hostInfo` obtain the properties and return the key of each properties

**Comment by AlbericByte** [comment]
@abhishekagarwal87 and @cryptoe may i continue to work on this, seems there is no response from 3 months ago.
and i have a similar pr : https://github.com/apache/druid/pull/16024#issuecomment-1976131021
cc:@mustajibmk

**Comment by abhishekagarwal87** [comment]
Go ahead please

**Comment by FrankChen021** [comment]
@AlbericByte let's continue

**Comment by FrankChen021** [comment]
@abhishekagarwal87 Let's shift to  #16024 since @AlbericByte has also finished the work and I see all the CI checks are green.

Still thanks the work from @mustajibmk

**Pull request closed without merging**